### DISCUSSION TITLE
Why does `spdlog::register_logger()` throw an exception?

### DISCUSSION THREAD

**Discussion by wolf-chesty** [gabime/spdlog#3226] [General]
Hey,

For my application I need to be able to dynamically change the logging directory. Looking at the code I've noticed that the function `spdlog::set_default_logger()` can be invoked multiple times and will replace the default logger being used by spdlog in a manner that is thread safe.

`spdlog::register_logger()` doesn't behave the same way. The way spdlog works currently actually requires that we invoke 'spdlog::drop()` before replacing the logging target. Like so:
```
spdlog::drop("logger_name");
// Accessing the log sink here can crash the application
spdlog::register_logger(new_logger);
```

We aren't able to replace a non-default logging sink, only the default logger can be replaced in a thread-safe manner. Is there a reason for this?

Thanks,

**Comment by tt4g** [comment]
Loggers with the same name cannot be registered in spdlog's logger registry.
You must use `spdlog::drop()` to delete loggers with the same name.

The reason why `spdlog::set_default_logger()` automatically replaces loggers is so that there is no moment when a logger accessed by the global logging API, such as `spdlog::info()`, is disabled.

You are using an API that is not designed to replace sinks.
Instead of replacing loggers, use dist_sink, which supports sink replacement: https://github.com/gabime/spdlog/wiki/4.-Sinks

You can use `spdlog::sinks::dist_sink_mt` and call `spdlog::dist_sink::set_sinks()` to replace sinks.

https://github.com/gabime/spdlog/blob/e593f6695c6065e6b345fe2862f04a519ed484e0/include/spdlog/sinks/dist_sink.h#L42-L45
